### PR TITLE
Agolubev #17338 add output stream source graph stage based

### PR DESCRIPTION
**Pull request by agolubev**
This is for ticket #17338. 
Current brunch implements OuputStreamSource with GraphStageWithMaterializedValue as @drewhk suggested (InsputStreamSink stays unchanged). First version is in PR #17995 and based on ActorPublisher/ActorSubscriber.
Please let me know if this approach is good enough to fix the ticket.
Currently one test javadsl/scaladsl is failed as in javadsl new apply method is missed (had to introduce scaladsl.SourceApply.apply with builder materializer type)

### CODE REVIEW

**Comment by ktoso** [comment]
OK TO TEST

File: akka-stream-tests/src/test/java/akka/stream/io/OutputStreamSourceTest.java [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+        final JavaTestKit probe = new JavaTestKit(system);
+
+        final Source<ByteString, OutputStream> source = OutputStreamSource.create(timeout);
+        final OutputStream s = source.to(Sink.foreach(new Procedure<ByteString>() {
+            public void apply(ByteString elem) {
+                probe.getRef().tell(elem, ActorRef.noSender());
+            }
+        })).run(materializer);
+
+        s.write("a".getBytes());
```
Comment (viktorklang) [inline akka-stream-tests/src/test/java/akka/stream/io/OutputStreamSourceTest.java]: What encoding?

Comment (agolubev) [inline akka-stream-tests/src/test/java/akka/stream/io/OutputStreamSourceTest.java]: Hm, OutputStream works with byte array, OutputStreamSource with ByteString which is wrapper for byte array as well. 
To be fully consistent with converting string to bytes and then string to ByteString will use getBytes instead of fromString below.

**Comment by agolubev** [comment]
Let me summarize my experience for GraphStage(WithMaterializedValue) API (main code is placed in ```OutputStreamSourceStage```), I believe some things are missed and some I just did not find:
1) ```GraphStageWithMaterializedValue``` is a good way of dealing with actors underneath with help of some wrapper. So we are thread safe with actor underneath but we don’t dealing with sending messages just calling methods
2) Source module boilerplate code is not needed anymore and can be removed
3) As soon as data (byte arrays) is not going through actor messages (however it will go through stream as akka messages anyway) and because I can not use ```GraphStage``` with ```akka.stream.blocking-io-dispatcher``` I removed such dependency
4) Source.apply has builder with Mat as [Unit] - so need to introduce another factory method to have it as generic [Mat]
5) Code that’s creating Source from ```GraphStageWithMaterializedValue``` looks strange
```scala
 Source.apply { implicit b: FlowGraph.Builder[OutputStream] ⇒
      val source = b.add(new OutputStreamSourceStage(16, timeout))
      source.outlet
    }
```
6) ```GraphStage``` is useful for some flow transformations and supposed to be used with Builder and Graphs a lot. Now it’s uses only as layer to abstract over ```SubscriberActor```
7) Did not find a way to get actor settings from within ```GraphStage```. It supposed to be read buffer size max from there.
8) Also was trying to implement blocking call (flush for example) with just ```() => Future[Unit]``` from ```OutputStreamAdapter``` to ```GraphStage``` (```GraphStage``` does not have API for calls from outside of reactive stream context). Future[Unit] will be finished when all elements sent downstream on stream is canceled.
```scala
val logic = new GraphStageLogic {
     def wakeUp(msg: AdapterToStageMessage): Future[Unit] = {
        implicit val ex = interpreter.materializer.executionContext
        val p = Promise[Unit]
        Future(message.invoke((msg, p)))
        p.future
      }
      ......
}

(logic, new OutputStreamAdapter(dataQueue, downstreamStatus, logic.wakeUp, timeout))
```
The following code gets me warning: "reflective access of structural type member value message should be enabled"
Did not find the simple way to fix it (believe this will impact performance in a bad way)

Also few points for general approach
1) The main purpose for using ```BlockingQueue``` was to reduce number of big data messages with byte arrays. However downstream will send bytes in messages anyway. So ```OutputStreamSource``` suppose to be just another node that transferring data in messages. My point is there will be performance impact anyway just because of nature of standard stream components. So this is performance of one stream element versus consistency (with other IO Sinks/Sources) choice
2) Removing ask’s with block calls to ```BlockingQueues```... It’s not always a good idea to have ```BlockingQueue``` with one element (like for status). Also we should have timeouts anyway to support fail fast idea. So will have some kind of timeouts anyway.
3) In previous implementation Source was materialized to ```(OutputStream, Future[Long])``` with number of written bytes when it’s completed. This was done alike other ```InputStreamSource``` and ```OutputStreamSink```. Those are behave as pipes and have lack of visibility so that was the reason Future[Long] was introduced. ```OutputStreamSource```, on the other side, requires explicit writing calls so developer will have pretty much visibility on what’s going on there. With this in mind I removed ```Future[Long]``` from ```OutputStreamSource```

To sum up, this is third general refactoring (or reimplementation) of this functionality. I believe we should have pretty good vision on implementation options for now. So let me know what approach I need to stick to to finalize this fix (hope that last one is good enough :) ).

File: akka-stream/src/main/scala/akka/stream/io/OutputStreamSource.scala [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+  import scala.language.postfixOps
+
+  /**
+   * Creates a synchronous (Java 6 compatible) Source.
+   *
+   * It materializes a [[Future]] containing the number of bytes written to OutputStream upon completion and
+   * [[OutputStream]] to interacting with reactive stream.
+   */
+  def apply(timeout: FiniteDuration = 5.seconds): Source[ByteString, OutputStream] = {
+    Source.apply2 { implicit b: FlowGraph.Builder[OutputStream] ⇒
```
Comment (drewhk) [inline akka-stream/src/main/scala/akka/stream/io/OutputStreamSource.scala]: Since `OutputStreamSourceStage` has the right shape already, all you need is:

    Source.wrap(new OutputStreamSourceStage(...))

Comment (drewhk) [inline akka-stream/src/main/scala/akka/stream/io/OutputStreamSource.scala]: Which reads roughly like "add the nice DSL to the thing whch is already a Source-linke thing"

File: akka-stream-tests/src/test/java/akka/stream/io/OutputStreamSourceTest.java (hunk 1 above)
Comment (viktorklang) [inline akka-stream-tests/src/test/java/akka/stream/io/OutputStreamSourceTest.java]: String.getBytes uses the platform encoding whereas ByteString.fromString uses UTF-8

File: akka-stream/src/main/scala/akka/stream/impl/io/OutputStreamPublisher.scala [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+    } else throw new IOException("OutputStream is closed")
+
+  @scala.throws(classOf[IOException])
+  override def write(b: Int) = {
+    sendData(ByteString(b))
+  }
+
+  @scala.throws(classOf[IOException])
+  override def write(b: Array[Byte], off: Int, len: Int) = {
+    sendData(ByteString(b).drop(off))
```
Comment (viktorklang) [inline akka-stream/src/main/scala/akka/stream/impl/io/OutputStreamPublisher.scala]: Most probably cheaper to do: ByteString(b, off, len)

File: akka-stream-tests/src/test/java/akka/stream/io/InputStreamSinkTest.java [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+    public void mustReadEventViaInputStream() throws Exception {
+        final FiniteDuration timeout = FiniteDuration.create(300, TimeUnit.MILLISECONDS);
+
+        final Sink<ByteString, Pair<InputStream,Future<Long>>> sink = InputStreamSink.create(timeout);
+        final List<ByteString> list = Collections.singletonList(ByteString.fromString("a"));
+        final Pair<InputStream, Future<Long>> p = Source.from(list).runWith(sink, materializer);
+
+        byte[] a = new byte[1];
+        p.first().read(a);
+        assertTrue(Arrays.equals("a".getBytes(), a));
```
Comment (viktorklang) [inline akka-stream-tests/src/test/java/akka/stream/io/InputStreamSinkTest.java]: Put a specific character encoding for all invocations of String.getBytes(), otherwise it'll use platform default.

Comment (viktorklang) [inline akka-stream-tests/src/test/java/akka/stream/io/InputStreamSinkTest.java]: https://docs.oracle.com/javase/7/docs/api/java/lang/String.html#getBytes()

File: akka-stream-tests/src/test/java/akka/stream/io/InputStreamSinkTest.java [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+    @ClassRule
+    public static AkkaJUnitActorSystemResource actorSystemResource = new AkkaJUnitActorSystemResource("InputStreamSink",
+            Utils.UnboundedMailboxConfig());
+    @Test
+    public void mustReadEventViaInputStream() throws Exception {
+        final FiniteDuration timeout = FiniteDuration.create(300, TimeUnit.MILLISECONDS);
+
+        final Sink<ByteString, Pair<InputStream,Future<Long>>> sink = InputStreamSink.create(timeout);
+        final List<ByteString> list = Collections.singletonList(ByteString.fromString("a"));
+        final Pair<InputStream, Future<Long>> p = Source.from(list).runWith(sink, materializer);
```
Comment (viktorklang) [inline akka-stream-tests/src/test/java/akka/stream/io/InputStreamSinkTest.java]: Iso using List here you can use Source.single(ByteString(…))

File: akka-stream-tests/src/test/java/akka/stream/io/OutputStreamSourceTest.java (hunk 1 above)
Comment (agolubev) [inline akka-stream-tests/src/test/java/akka/stream/io/OutputStreamSourceTest.java]: yeah, I'll fix this

File: akka-stream/src/main/scala/akka/stream/io/OutputStreamSource.scala (hunk 2 above)
Comment (agolubev) [inline akka-stream/src/main/scala/akka/stream/io/OutputStreamSource.scala]: Thanks, that's what I missed

File: akka-stream-tests/src/test/java/akka/stream/io/InputStreamSinkTest.java (hunk 4 above)
Comment (agolubev) [inline akka-stream-tests/src/test/java/akka/stream/io/InputStreamSinkTest.java]: it's using file.encoding system property. I got it. Will change it to have UTC-8

Comment (viktorklang) [inline akka-stream-tests/src/test/java/akka/stream/io/InputStreamSinkTest.java]: Make that UTF-8 :)

File: akka-stream-tests/src/test/java/akka/stream/io/OutputStreamSourceTest.java (hunk 1 above)
Comment (viktorklang) [inline akka-stream-tests/src/test/java/akka/stream/io/OutputStreamSourceTest.java]: awesome, thanks!

File: akka-stream-tests/src/test/scala/akka/stream/io/InputStreamSinkSpec.scala [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+
+class InputStreamSinkSpec extends AkkaSpec(UnboundedMailboxConfig) {
+  import system.dispatcher
+
+  val settings = ActorMaterializerSettings(system).withDispatcher("akka.actor.default-dispatcher")
+  implicit val materializer = ActorMaterializer(settings)
+
+  val timeout = 300.milliseconds
+  val byteString = ByteString("abc")
+  val byteArray = "abc".toArray
```
Comment (viktorklang) [inline akka-stream-tests/src/test/scala/akka/stream/io/InputStreamSinkSpec.scala]: java.lang.String does not have a toArray method, so I assume it is added by Scala, and I don't know what char encoding it uses. I'd recommend doing: ByteString.fromString("abc", "UTF-8").toArray

Comment (viktorklang) [inline akka-stream-tests/src/test/scala/akka/stream/io/InputStreamSinkSpec.scala]: One idea would be to not use Strings for the byte arrays, since the values themselves are not interesting.
instead, it would be easier to generate random byte-arrays of a given length and compare to those instead.

File: akka-stream-tests/src/test/scala/akka/stream/io/OutputStreamSourceSpec.scala [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+  import system.dispatcher
+
+  val settings = ActorMaterializerSettings(system)
+  implicit val materializer = ActorMaterializer(settings)
+
+  val timeout = 300.milliseconds
+  val bytes = "abc".getBytes
+  val byteString = ByteString("abc")
+
+  def expectBlocked[T](f: Future[T]) =
```
Comment (viktorklang) [inline akka-stream-tests/src/test/scala/akka/stream/io/OutputStreamSourceSpec.scala]: blocked?

File: akka-stream-tests/src/test/scala/akka/stream/io/OutputStreamSourceSpec.scala [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+      val (outputStream, probe) = OutputStreamSource().toMat(TestSink.probe[ByteString])(Keep.both).run
+
+      probe.expectSubscription()
+      outputStream.close()
+      probe.expectComplete()
+      the[Exception] thrownBy outputStream.write(bytes) shouldBe a[IOException]
+    }
+
+    "throw IOException when writing to the stream after the subscriber has cancelled the reactive stream" in assertAllStagesStopped {
+      pending
```
Comment (viktorklang) [inline akka-stream-tests/src/test/scala/akka/stream/io/OutputStreamSourceSpec.scala]: Why pending?

File: akka-stream-tests/src/test/scala/akka/stream/scaladsl/GraphBackedFlowSpec.scala [hunk 9]
```diff
@@ -141,7 +141,7 @@ class GraphFlowSpec extends AkkaSpec {
       "work with a Sink when having KeyedSource inside" in {
         val probe = TestSubscriber.manualProbe[Int]()
 
-        val source = Source.apply(Source.subscriber[Int]) { implicit b ⇒
+        val source = Source(Source.subscriber[Int]) { implicit b ⇒
```
Comment (viktorklang) [inline akka-stream-tests/src/test/scala/akka/stream/scaladsl/GraphBackedFlowSpec.scala]: :)

File: akka-stream/src/main/boilerplate/akka/stream/scaladsl/SourceApply.scala.template [hunk 10]
```diff
@@ -7,6 +7,12 @@ import akka.stream.{ Outlet, Shape, Graph }
 
 trait SourceApply {
 
+  def apply2[Out,Mat](buildBlock: FlowGraph.Builder[Mat] => Outlet[Out]): Source[Out, Mat] = {
```
Comment (viktorklang) [inline akka-stream/src/main/boilerplate/akka/stream/scaladsl/SourceApply.scala.template]: Why is this needed?

File: akka-stream/src/main/scala/akka/stream/impl/io/IOSinks.scala [hunk 11]
```diff
@@ -66,3 +74,163 @@ private[akka] final class OutputStreamSink(createOutput: () ⇒ OutputStream, va
...
+/**
+ * INTERNAL API
+ * Creates simple synchronous (Java 6 compatible) Sink which provide all incoming element on read request via
+ * InputStream.
+ */
+private[akka] class InputStreamSink(val timeout: FiniteDuration,
+                                    val attributes: Attributes, shape: SinkShape[ByteString])
+  extends SinkModule[ByteString, (InputStream, Future[Long])](shape) {
+
+  protected def getSubscriber(sharedBuffer: BlockingQueue[ByteString]) =
```
Comment (viktorklang) [inline akka-stream/src/main/scala/akka/stream/impl/io/IOSinks.scala]: We always use explicit return type parameters, would you mind adding them? Thank you! :)

File: akka-stream-tests/src/test/scala/akka/stream/io/OutputStreamSourceSpec.scala (hunk 7 above)
Comment (agolubev) [inline akka-stream-tests/src/test/scala/akka/stream/io/OutputStreamSourceSpec.scala]: flush and close suppose to be blocked until all data is pulled from downstream or stream is canceled.
Should I rename method? or you have other concern?

File: akka-stream-tests/src/test/scala/akka/stream/io/OutputStreamSourceSpec.scala (hunk 8 above)
Comment (agolubev) [inline akka-stream-tests/src/test/scala/akka/stream/io/OutputStreamSourceSpec.scala]: This test case requires some synchronization. We need to know when stage received "canceled" message from downstream. I created test version of SourceState for previous implementation - will figure something out with GraphStage as well. Until the whole approach is not approved would like to keep it pending

File: akka-stream-tests/src/test/scala/akka/stream/scaladsl/GraphBackedFlowSpec.scala (hunk 9 above)
Comment (agolubev) [inline akka-stream-tests/src/test/scala/akka/stream/scaladsl/GraphBackedFlowSpec.scala]: :) will change this back as change is outside of ticket scope - (until you want me to have it stayed)

File: akka-stream-tests/src/test/scala/akka/stream/io/OutputStreamSourceSpec.scala (hunk 7 above)
Comment (viktorklang) [inline akka-stream-tests/src/test/scala/akka/stream/io/OutputStreamSourceSpec.scala]: expectTimeout? (and should perhaps take a timeout parameter)

File: akka-stream/src/main/boilerplate/akka/stream/scaladsl/SourceApply.scala.template (hunk 10 above)
Comment (agolubev) [inline akka-stream/src/main/boilerplate/akka/stream/scaladsl/SourceApply.scala.template]: Initially I looked for apply with [Out,Mat] instead of [Out,Unit] but @drewhk gave me a hint that Source.wrap will work here

File: akka-stream-tests/src/test/scala/akka/stream/scaladsl/GraphBackedFlowSpec.scala (hunk 9 above)
Comment (viktorklang) [inline akka-stream-tests/src/test/scala/akka/stream/scaladsl/GraphBackedFlowSpec.scala]: Why not `Source.subscriber[Int]`?

File: akka-stream/src/main/boilerplate/akka/stream/scaladsl/SourceApply.scala.template (hunk 10 above)
Comment (viktorklang) [inline akka-stream/src/main/boilerplate/akka/stream/scaladsl/SourceApply.scala.template]: Ah, yes, good point. Source.wrap will do it!

File: akka-stream-tests/src/test/scala/akka/stream/io/OutputStreamSourceSpec.scala [hunk 12]
```diff
@@ -0,0 +1,135 @@
...
+
+      outputStream.close()
+      probe.expectComplete()
+    }
+
+    "block writes when buffer is full" in assertAllStagesStopped {
+      val (outputStream, probe) = OutputStreamSource().toMat(TestSink.probe[ByteString])(Keep.both).run
+      val s = probe.expectSubscription()
+
+      //default max-input-buffer-size = 16
```
Comment (viktorklang) [inline akka-stream-tests/src/test/scala/akka/stream/io/OutputStreamSourceSpec.scala]: Instead of this implicit dependency, how about manually specifying it as a setting in the graph?

File: akka-stream-tests/src/test/scala/akka/stream/io/OutputStreamSourceSpec.scala (hunk 8 above)
Comment (viktorklang) [inline akka-stream-tests/src/test/scala/akka/stream/io/OutputStreamSourceSpec.scala]: ok

File: akka-stream/src/main/scala/akka/stream/impl/io/IOSinks.scala (hunk 11 above)
Comment (agolubev) [inline akka-stream/src/main/scala/akka/stream/impl/io/IOSinks.scala]: Will fix this. Please don't review InputStreamSink/InputStreamSubscriber part as this is still actor based (previous version). Now I'm trying to implement with using GraphStage as @drewhk  suggested

File: akka-stream-tests/src/test/scala/akka/stream/io/OutputStreamSourceSpec.scala (hunk 7 above)
Comment (agolubev) [inline akka-stream-tests/src/test/scala/akka/stream/io/OutputStreamSourceSpec.scala]: Yes, it makes sense.

File: akka-stream-tests/src/test/scala/akka/stream/io/OutputStreamSourceSpec.scala (hunk 12 above)
Comment (agolubev) [inline akka-stream-tests/src/test/scala/akka/stream/io/OutputStreamSourceSpec.scala]: Ok, will it be like GraphStage.withAttributes fro set and then GraphStage.module.attributes to read?

File: akka-stream-tests/src/test/scala/akka/stream/scaladsl/GraphBackedFlowSpec.scala (hunk 9 above)
Comment (agolubev) [inline akka-stream-tests/src/test/scala/akka/stream/scaladsl/GraphBackedFlowSpec.scala]: Yeah this is example I found and went the wrong way - will simplify it to Source.subscriber[Int]

File: akka-stream-tests/src/test/scala/akka/stream/io/OutputStreamSourceSpec.scala (hunk 12 above)
Comment (agolubev) [inline akka-stream-tests/src/test/scala/akka/stream/io/OutputStreamSourceSpec.scala]: Ok, found example       
```scala
val p = SynchronousFileSource(testFile, chunkSize)
        .withAttributes(bufferAttributes)
        .runWith(Sink.publisher)
```

Comment (ktoso) [inline akka-stream-tests/src/test/scala/akka/stream/io/OutputStreamSourceSpec.scala]: yup exactly

**Comment by agolubev** [comment]
@viktorklang, implemented InputStreamSink as GraphStage as well, did rebase.
Now the whole feature is available for code review

**Comment by agolubev** [comment]
Zip spec happy scenario failed for some reason during test run.

File: akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala [hunk 13]
```diff
@@ -61,6 +61,24 @@ abstract class GraphStage[S <: Shape] extends GraphStageWithMaterializedValue[S,
   def createLogic: GraphStageLogic
 }
 
+/**
+ * A SourceStage represents a reusable graph stream processing stage. A SourceStage consists of a [[akka.stream.Shape]] which describes
+ * its output port.
+ */
+abstract class SourceStage[Out, M](name: String) extends GraphStageWithMaterializedValue[SourceShape[Out], M] {
```
Comment (drewhk) [inline akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala]: I would add convenience methods, `push(elem) ` == `push(out, elem)` and similarly for `pull()`

**Comment by drewhk** [comment]
PLS BUILD

**Comment by drewhk** [comment]
LGTM. This will need a documentation section in the *Integration* section of the reference docs, but that can be another PR (and can be done by us). Can you open a ticket for that?

**Comment by agolubev** [comment]
Ok, will crate additional ticket

File: akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala (hunk 13 above)
Comment (agolubev) [inline akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala]: @drewhk, I tried it in few different ways and the most simple was:
```scala
abstract class SourceStage[Out, M](name: String) extends GraphStageWithMaterializedValue[SourceShape[Out], M] {
  def push(elem:Out) = createLogicAndMaterializedValue._1.push(out,elem)
```
which is calling createLogicAndMaterializedValue and creating GraphStageLogic another time. As far as I understand this call must be made only by materializer. Can we skip this for now (I don't see easy way of implementing this)?

Comment (drewhk) [inline akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala]: ah no, that method should be on a subclass of GraphStageLogic, but nevermind we can do this later.

Comment (agolubev) [inline akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala]: ok, will do rebase with squash than

Comment (drewhk) [inline akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala]: Thanks, great work!

**Comment by agolubev** [comment]
@drewhk,  PR is rebased

**Comment by drewhk** [comment]
In it goes.
